src/dbnode/persist/fs/registry.go

```diff
@@ -0,0 +1,335 @@
+	if err != nil {
+		return nil, err
+	}
+	return entry.OpenFile(flag, perm)
+}
+
+func (r *fileRegistry) Remove(filePath string) error {
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return err
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	status, ok := r.entries[pathAbs]
+	if !ok {
+		// Just fine, no files open.
+		return os.Remove(pathAbs)
+	}
+
+	// Delete from the registry.
+	delete(r.entries, pathAbs)
+
+	// Notify open file holders that this file is being removed
+	// and perform a delayed remove which will delete the file
+	// after all references are released.
+	go func() {
+		err := status.NotifyAndRemove()
+		if err != nil {
+			r.logger.Error("unable to delayed remove file",
+				zap.String("path", pathAbs), zap.Error(err))
+		}
+	}()
+
+	return nil
+}
+
+func (r *fileRegistry) newFileStatusOrExisting(filePath string) (*fileStatus, error) {
+	pathAbs, err := filepath.Abs(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	r.RLock()
+	status, ok := r.entries[pathAbs]
+	r.RUnlock()
+
+	if ok {
+		return status, nil
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	status, ok = r.entries[pathAbs]
+	if !ok {
+		status = newFileStatus(pathAbs, r.clockOpts.NowFn(), r.logger)
+		r.entries[pathAbs] = status
+	}
+
+	return status, nil
+}
+
+type fileStatus struct {
+	sync.RWMutex
+	status    FileStatus
+	pathAbs   string
+	watchable watch.Watchable
+	refs      []*fileRef
+
+	nowFn clock.NowFn
+
+	logger *zap.Logger
+}
+
+func newFileStatus(pathAbs string, nowFn clock.NowFn, logger *zap.Logger) *fileStatus {
+	f := &fileStatus{
+		status: FileStatus{
+			Type: AvailableFileStatusType,
+		},
+		pathAbs: pathAbs,
+		nowFn:   nowFn,
+		logger:  logger,
+	}
+	watchable := watch.NewWatchable()
+	watchable.Update(f.status)
+	f.watchable = watchable
+	return f
+}
+
+func (f *fileStatus) newFileRef(ref *os.File) *fileRef {
+	fileRef := newFileRef(f, ref)
+	f.Lock()
+	f.refs = append(f.refs, fileRef)
+	f.Unlock()
+	return fileRef
+}
+
+func (f *fileStatus) removeFileRef(elem *fileRef) error {
+	f.Lock()
+	defer f.Unlock()
+
+	if len(f.refs) == 0 {
+		return fmt.Errorf("remove file ref not found for: %v", elem.PathAbs())
+	}
+
+	refs := make([]*fileRef, 0, len(f.refs)-1)
+	for _, ptr := range f.refs {
+		if ptr != elem {
+			refs = append(refs, ptr)
+		}
+	}
+	if len(refs) != len(f.refs)-1 {
+		return fmt.Errorf("remove file ref not found for: %v", elem.PathAbs())
+	}
+
+	f.refs = refs
+	return nil
+}
+
+func (f *fileStatus) Status() FileStatus {
+	f.RLock()
+	v := f.status
+	f.RUnlock()
+	return v
+}
+
+func (f *fileStatus) NotifyAndRemove() error {
+	f.setStatus(FileStatus{Type: RemovingFileStatusType})
+	defer func() {
+		f.setStatus(FileStatus{Type: RemovedFileStatusType})
+	}()
+
+	timeout := f.nowFn().Add(notifyAndRemoveTimeout)
+	for f.nowFn().Before(timeout) && f.numRefs() > 0 {
+		time.Sleep(notifyAndRemoveCheckInterval)
+	}
+
+	if f.numRefs() > 0 {
+		f.logger.Error("timed out for delayed remove of file",
```

Comment:
do you need to return here too?

Reply:
or is the expected semantic that the file will be removed and zombie handles are the watchers responsibility now?

Reply:
I was thinking we just delete anyway, thoughts?  Otherwise a server restart wouldn't fix the issue.

Reply:
Could we just do this all sync and return an error here? Then whatever process encountered the issue can at least try again

Reply:
Well the NotifyAndRemove can still be done async in parallel, but then we would wait for them to all complete in `Remove()` and if they timedout they would return errors instead of just deleting anyways because then we can try again from the callers perspective